Revise and condense the plugin tutorial (#103)
Limnoria/Limnoria-doc: develop/plugin_tutorial.rst

the ``Class = Random`` statement at the end of the file). It is helpful to fill
in the plugin docstring with some more details on how to actually use the plugin
too: this info can be shown on a live bot using the
``plugin help <plugin name>`` command.

::

    class Random(callbacks.Plugin):
        """This plugin contains commands relating to random numbers, including random sampling from a list and a simple dice roller."""

        def __init__(self, irc):
            # Make sure to call the superclass' constructor when you define a custom one
            super().__init__(irc)
            self.rng = random.Random()  # create our rng
            self.rng.seed() # automatically seeds with current time

For this sample plugin, we define a custom constructor (``__init__``) that
instantiates a random number generator instance and pre-seeds it. This isn't
technically necessary for Python's ``random`` module, but this helps outline
how to write a similar constructor. Notice in particular how you must pass in
the ``irc`` argument in addition to ``self``.

.. note::
    TODO(jlu): semantically, what does ``irc`` refer to? Most plugins don't

>     actually reference it on load time.

Reviewer: `Irc` object of the first network the bot connects to (on startup) or the network from which the plugin is loaded (if `@load`)

Basic command handler
---------------------

Our first command definition can immediately follow::

    # dummy comment to indent the below code consistently
        @wrap
        def random(self, irc, msg, args):
            """takes no arguments

            Returns the next random number from the random number generator.
            """
            irc.reply(str(self.rng.random()))

.. note::
    All functions used as commands must have an all lowercase name.

A command function taking in no arguments from IRC will still require 4
arguments; they are as follows:

- ``self``: refers to the class instance. It is common to keep local state
  for the plugin as instance variables within the plugin class.
- ``irc``: refers to the IRC network instance the command was called on

> - ``msg``: a :ref:`supybot.ircmsgs <supybot-ircmsgs>` instance; refers to the
>   IRC message that triggered this command.

Reviewer: ```suggestion
- ``msg``: a :class:`supybot.ircmsgs.IrcMsg` instance; refers to the
  IRC message that triggered this command.
```

arguments; they are as follows:

- ``self``: refers to the class instance. It is common to keep local state
  for the plugin as instance variables within the plugin class.
- ``irc``: refers to the IRC network instance the command was called on
- ``msg``: a :class:`supybot.ircmsgs.IrcMsg` instance; refers to the
  IRC message that triggered this command.
- ``args``: a raw list of remaining unconverted arguments; new plugins that
  use :ref:`@wrap <using-wrap>` for automatic argument type conversion should
  never need to interact with ``args`` directly.

The function docstring is expected to be in a particular format. First, the very
first line dictates the argument list to be displayed when someone calls the
``help`` command on this command (i.e., ``help random``). Then, leave a blank
line and start the actual help string for the function. Don't worry about the
fact that it's tabbed in or anything like that, as the help command normalizes
it to make it look nice. This part should be fairly brief but sufficient to
explain the function and what (if any) arguments it requires. Remember that this
should fit in one IRC message which is typically around a 450 character limit.

> The :py:meth:`irc.reply <supybot.callbacks.NestedCommandsIrcProxy.reply>` call
> is a bit of magic: it issues a reply the same place as the message that
> triggered the command. i.e. this may be in a channel or in a private
> conversation with the bot.

Reviewer: ```suggestion
The :py:meth:`irc.reply <supybot.callbacks.ReplyIrcProxy.reply>` call
is a bit of magic: it issues a reply the same place as the message that
triggered the command. i.e. this may be in a channel or in a private
conversation with the bot.
```

`ReplyIrcProxy` is the parent class; and IIRC `NestedCommandsIrcProxy` is only passed to nested commands. Regardless, `ReplyIrcProxy`'s `reply` method is better documented

sample of arbitrary size from a list provided by the user:

.. code-block::
    :dedent: 0

        @wrap(['int', many('anything')])
        def sample(self, irc, msg, args, n, items):
            """<number of items> <item1> [<item2> ...]

            Returns a sample of the <number of items> taken from the remaining
            arguments.  <number of items> must be less than the number
            of arguments given.
            """
            if n > len(items):
                # Calling irc.error with Raise=True is an alternative early return
                irc.error('<number of items> must be less than the number '
                          'of arguments.', Raise=True)
            sample = self.rng.sample(items, n)
            sample.sort()
            irc.reply(utils.str.commaAndify(sample))

The important thing to note is that list type arguments are rolled into one
parameter in the command function by the ``many`` filter. Similar "multiplicity"
handlers are documented :ref:`here <wrap-multiplicity-handlers>`.

We also update the docstring to use the ``[]`` convention when surrounding
optional arguments.

> For this function's body,
> :py:meth:`irc.error <supybot.callbacks.NestedCommandsIrcProxy.error>`
> is like
> :py:meth:`irc.replySuccess <supybot.callbacks.NestedCommandsIrcProxy.replySuccess>`
> but for error messages. We prefer using this instead of ``irc.reply`` for error
> signaling because its behaviour can be configured specially. For example, you
> can force all errors to go in private by setting the ``reply.error.inPrivate``
> option, and this can help reduce noise on a busy channel.
> Also, ``irc.error()`` with no text will return a generic error message
> configured in ``supybot.replies.error``, but this is not a valid call to
> :py:meth:`irc.reply <supybot.callbacks.NestedCommandsIrcProxy.reply>`.

Reviewer: ```suggestion
For this function's body,
:py:meth:`irc.error <supybot.callbacks.ReplyIrcProxy.error>`
is like
:py:meth:`irc.replySuccess <supybot.callbacks.ReplyIrcProxy.replySuccess>`
but for error messages. We prefer using this instead of ``irc.reply`` for error
signaling because its behaviour can be configured specially. For example, you
can force all errors to go in private by setting the ``reply.error.inPrivate``
option, and this can help reduce noise on a busy channel.
Also, ``irc.error()`` with no text will return a generic error message
configured in ``supybot.replies.error``, but this is not a valid call to
:py:meth:`irc.reply <supybot.callbacks.ReplyIrcProxy.reply>`.
```

ditto

``utils.str.commaAndify`` is a simple helper that takes a list of strings
and turns it into "item1, item2, item3, item4, and item5" for an arbitrary
length. Limnoria has accumulated many such helpers in its lifetime, many of
which are described in the :ref:`Using Utils <using-utils>` page.

Command handler with optional arguments
---------------------------------------
Now for the last command that we will add to our plugin.py. This ``diceroll``
command will allow the bot users to roll an arbitrary n-sided die, with n
defaulting to 6:

.. code-block::
    :dedent: 0

        @wrap([additional(('int', 'number of sides'), 6)])
        def diceroll(self, irc, msg, args, n):
            """[<number of sides>]

            Rolls a die with <number of sides> sides.  The default number of sides
            is 6.
            """
            s = 'rolls a %s' % self.rng.randrange(1, n)
            irc.reply(s, action=True)

> The only new thing described here is that ``irc.reply(..., action=True)`` makes
> the bot perform a `/me`. There are some other flags described in the
> :py:meth:`irc.reply <supybot.callbacks.NestedCommandsIrcProxy.reply>`
> documentation too: common ones include ``private=True``, which
> forces a private message, and ``notice=True``, which forces the reply to use
> NOTICE instead of PRIVMSG.

Reviewer: ```suggestion
The only new thing described here is that ``irc.reply(..., action=True)`` makes
the bot perform a `/me`. There are some other flags described in the
:py:meth:`irc.reply <supybot.callbacks.ReplyIrcProxy.reply>`
documentation too: common ones include ``private=True``, which
forces a private message, and ``notice=True``, which forces the reply to use
NOTICE instead of PRIVMSG.
```
